core/src/main/scala/org/apache/spark/util/collection/ReferenceCounter.scala

```diff
@@ -0,0 +1,122 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.util.collection
+
+import scala.collection.JavaConverters._
+
+import com.google.common.cache.{CacheBuilder, CacheLoader}
+import com.google.common.collect.ConcurrentHashMultiset
+
+import org.apache.spark.TaskContext
+
+/**
+ * Thread-safe collection for maintaining both global and per-task reference counts for objects.
+ */
+private[spark] class ReferenceCounter[T] {
```

Comment:
I need to maintain global counts per each object as well as counts for each task (in order to automatically decrement the global counts when tasks finish) (I'm working on adding the `releaseAllReferencesForTask()` call to the task completion cleanup code).

If I stored the global count per block inside of the `BlockInfo` class, then I'd still need a mechanism to count the references per task. If the counts for each task were stored in `BlockInfo` then I'd have to loop over the `BlockInfo` list on task completion in order to clear those counts, or would have to maintain the counts separately. As a result, it made sense to me to keep both types of counts in close proximity like this.